### PR TITLE
Expand Zod validation to include `fallback` w/ optional inputs & fix `AbiError` validation 

### PR DESCRIPTION
**Pull request by Bind**
:wave: Hi there, was using this library in project of mine and bumped into some validation errors when comparing against contract ABI's in the wild. Some of these changes might be outside the scope of `abitype` as they might not strictly adhere to the [Contract ABI Spec](https://docs.soliditylang.org/en/latest/abi-spec.html) but I figured I'd open a PR and see what y'all think.  

## Description

- Mark `type: "fallback"` inputs as optional (supports legacy contracts like WETH)
- Update `AbiError` object, changing `type` from `z.literal("event")` to `z.literal("error")`  
- Moved `test-d` files to `test` (Happy to move them back, just wanted to run all test files while iterating!)
- Added tests to validate `Abi.parse` successfully evaluates for `test/abis`  

## Additional Information

- [x] I read the [contributing docs](/wagmi-dev/abitype/blob/main/.github/CONTRIBUTING.md) (if this is your first contribution)

Your ENS/address: doug.binder.eth

## TODO:
- [x] Potentially move `test-d` files back.
- [ ] Generate changeset if maintainers are interested in upstreaming.

### CODE REVIEW

File: src/zod/zod.ts [hunk 1]
```diff
@@ -137,7 +140,7 @@ export const AbiEvent = z.object({
 })
 
 export const AbiError = z.object({
-  type: z.literal('event'),
+  type: z.literal('error'),
```
Comment (tmm) [inline src/zod/zod.ts:143]: good catch

File: test/index.ts [hunk 2]
```diff
@@ -6,4 +6,6 @@ export {
   nounsAuctionHouseAbi,
   wagmiMintExampleAbi,
   writingEditionsFactoryAbi,
+  ensAbi,
+  customSolidityErrorsAbi,
```
Comment (tmm) [inline test/index.ts:10]: let's export `wethAbi` from here too and make sure any import of it comes from here.

File: src/zod/zod.test.ts [hunk 3]
```diff
@@ -1,5 +1,15 @@
...
 
+import {
+  customSolidityErrorsAbi,
+  ensRegistryWithFallbackAbi,
+  nestedTupleArrayAbi,
+  nounsAuctionHouseAbi,
+  wagmiMintExampleAbi,
+  wethAbi,
+  writingEditionsFactoryAbi,
+} from '../../test/abis'
```
Comment (tmm) [inline src/zod/zod.test.ts:11]:
```suggestion
} from '../../test'
```

File: src/zod/zod.test.ts [hunk 4]
```diff
@@ -1,5 +1,15 @@
 import { describe, expect, it } from 'vitest'
 
+import {
+  customSolidityErrorsAbi,
+  ensRegistryWithFallbackAbi,
```
Comment (tmm) [inline src/zod/zod.test.ts:5]: let's import `ensAbi` here and add to the test cases below too

File: src/zod/zod.test.ts [hunk 5]
```diff
@@ -34,7 +44,27 @@ describe('AbiSchema', () => {
...
+  })
+  it('successfully parses wethAbi', () => {
+    Abi.parse(wethAbi)
+  })
+  it('successfully parses writingEditionsFactoryAbi', () => {
+    Abi.parse(writingEditionsFactoryAbi)
+  })
+  it('successfully parses customSolidityErrorsAbi', () => {
+    Abi.parse(customSolidityErrorsAbi)
+  })
```
Comment (tmm) [inline src/zod/zod.test.ts:67]: let's group all these at the bottom in a `describe('ABI')` and chance the names to `it('nameOfAbi')`

**Review by tmm — COMMENTED** [review]
looking good! let's keep the `test-d.ts` files for now so [Vitest knows that they are type tests](https://vitest.dev/guide/testing-types.html) (maybe they will add better support for running a whole suite/we can write a `package.json` script)

**Comment by Bind** [comment]
Excellent! Addressed your feedback. Would you like me to generate a changeset for a minor version bump?

File: src/zod/zod.ts [hunk 6]
```diff
@@ -118,7 +118,10 @@ export const AbiFunction = z.preprocess(
         type: z.literal('constructor'),
         inputs: z.array(AbiParameter),
       }),
-      z.object({ type: z.literal('fallback'), inputs: z.tuple([]) }),
+      z.object({
+        type: z.literal('fallback'),
+        inputs: z.optional(z.tuple([])),
```
Comment (tmm) [inline src/zod/zod.ts:123]:
```suggestion
        inputs: z.tuple([]).optional(),
```

**Comment by tmm** [comment]
@Bind sounds great! let's go with `patch` since the following are required bug fixes that everyone needs to pick up.

> - Mark `type: "fallback"` inputs as optional (supports legacy contracts like WETH)
> - Update `AbiError` object, changing type from `z.literal("event")` to `z.literal("error")`

File: .changeset/dull-taxis-share.md [hunk 7]
```diff
@@ -0,0 +1,6 @@
+---
+'abitype': patch
+---
+
+- "fallback" type supports optional inputs (supports legacy contracts like WETH)
+- fix AbiError "type": change from z.literal("event") to z.literal("error")
```
Comment (tmm) [inline .changeset/dull-taxis-share.md:6]:
```suggestion
Fixed `abitype/zod` support for legacy contracts, like WETH, by making `inputs` optional for `type: 'fallback'` and `AbiError` `type` from `"event"` to `"error"`.
```

**Comment by tmm** [comment]
Thanks for contributing!

**Comment by Bind** [comment]
You made it easy! Thank you 🙌

**Pull request merged**